File: authentication-client/authentication-client-java/src/test/java/co/cask/cdap/security/authentication/client/basic/handlers/AuthenticationHandler.java
/*
 * Copyright 2014 Cask Data, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package co.cask.cdap.security.authentication.client.basic.handlers;

import co.cask.cdap.security.authentication.client.basic.BasicAuthenticationClientTest;
import com.google.common.base.Charsets;
import org.apache.commons.codec.binary.Base64;
import org.apache.commons.lang.StringUtils;
import org.apache.http.HttpException;
import org.apache.http.HttpHeaders;
import org.apache.http.HttpRequest;
import org.apache.http.HttpResponse;
import org.apache.http.HttpStatus;
import org.apache.http.RequestLine;
import org.apache.http.entity.StringEntity;
import org.apache.http.protocol.HttpContext;
import org.apache.http.protocol.HttpRequestHandler;

import java.io.IOException;
import javax.ws.rs.HttpMethod;
import javax.ws.rs.core.MediaType;

public class AuthenticationHandler implements HttpRequestHandler {
  @Override
  public void handle(HttpRequest httpRequest, HttpResponse httpResponse, HttpContext httpContext)
    throws HttpException, IOException {

    RequestLine requestLine = httpRequest.getRequestLine();
    String method = requestLine.getMethod();
    int statusCode;
    if (HttpMethod.GET.equals(method)) {
      String authHeaderVal = httpRequest.getFirstHeader(HttpHeaders.AUTHORIZATION).getValue();
      if (StringUtils.isNotEmpty(authHeaderVal)) {
        authHeaderVal = authHeaderVal.replace("Basic ", StringUtils.EMPTY);
        String credentialsStr = new String(Base64.decodeBase64(authHeaderVal), Charsets.UTF_8);
        String[] credentials = credentialsStr.split(":");
        String username = credentials[0];
        String password = credentials[1];
        if (BasicAuthenticationClientTest.USERNAME.equals(username) &&
          BasicAuthenticationClientTest.PASSWORD.equals(password)) {
          StringEntity entity = new StringEntity(createEntityBody(BasicAuthenticationClientTest.TOKEN,
                                                                  BasicAuthenticationClientTest.TOKEN_TYPE,
                                                                  BasicAuthenticationClientTest.TOKEN_LIFE_TIME));
          entity.setContentType(MediaType.APPLICATION_JSON);
          httpResponse.setEntity(entity);
          statusCode = HttpStatus.SC_OK;
        } else if (BasicAuthenticationClientTest.EMPTY_TOKEN_USERNAME.equals(username)) {
          StringEntity entity = new StringEntity(createEntityBody(StringUtils.EMPTY,
                                                                  BasicAuthenticationClientTest.TOKEN_TYPE,
                                                                  BasicAuthenticationClientTest.TOKEN_LIFE_TIME));
          entity.setContentType(MediaType.APPLICATION_JSON);
          httpResponse.setEntity(entity);
          statusCode = HttpStatus.SC_OK;
        } else {
          statusCode = HttpStatus.SC_UNAUTHORIZED;
        }
      } else {
        statusCode = HttpStatus.SC_BAD_REQUEST;
      }
    } else {
      statusCode = HttpStatus.SC_NOT_IMPLEMENTED;
    }
    httpResponse.setStatusCode(statusCode);
  }

  private static String createEntityBody(String value, String type, Long expiresIn) {
    return "{'access_token':'" + value + "','token_type':'" + type + "','expires_in':" + expiresIn + "}";
  }
}

File: authentication-client-python/tests/AuthenticationHandler.py
from BaseHTTPServer import BaseHTTPRequestHandler
import base64
import httplib
import BasicAuthenticationClientTest as bact


class AuthenticationHandler(BaseHTTPRequestHandler):
    AUTH_PORT = None
    AUTH_HOST = None

    @property
    def auth_host(self):
        return self.AUTH_HOST

    @auth_host.setter
    def auth_host(self, auth_host):
        self.AUTH_HOST = auth_host

    @property
    def auth_port(self):
        return self.AUTH_PORT

    @auth_host.setter
    def auth_port(self, auth_port):
        self.AUTH_PORT = auth_port

    def do_GET(self):
        request_ulr= self.requestline['/token']
        if not request_ulr :
            status_code = httplib.UNAUTHORIZED
            auth_uri = '{"auth_uri":["http://%s:%d/token"]}' % ( self.AUTH_HOST, self.AUTH_PORT)
            self.send_response(status_code)
            self.send_header("Content-type", "application/json")
            self.end_headers()
            self.wfile.write(auth_uri)
        else:
            request_counter = 0
            auth_header_val = self.headers['Authorization']
            if auth_header_val:
                auth_header_val.replace('Basic ','')
                credentials_str = base64.b64decode(auth_header_val)
                credentials = credentials_str.split(':', 1)
                username = credentials[0]
                password = credentials[1]
                if bact.USERNAME == username and bact.PASSWORD == password:
                    status_code = httplib.OK
                    self.wfile.write(self.create_resp_body(bact.TOKEN, bact.TOKEN_TYPE, bact.TOKEN_LIFE_TIME))
                elif bact.EMPTY_TOKEN_USERNAME == username:
                    self.wfile.write(self.create_resp_body('', bact.TOKEN_TYPE, bact.TOKEN_LIFE_TIME))
                    status_code = httplib.OK
                elif bact.EXPIRED_TOKEN_USERNAME == username:
                    if self.request_counter == 1:
                        resp = self.create_resp_body(bact.TOKEN, bact.TOKEN_TYPE, 5)
                    else:
                        resp = self.create_resp_body(bact.NEW_TOKEN, bact.TOKEN_TYPE, bact.TOKEN_LIFE_TIME)
                    self.wfile.write(resp)
                    status_code = httplib.OK
                else:
                    status_code = httplib.UNAUTHORIZED
            else:
                status_code = httplib.BAD_REQUEST

            self.send_response(status_code)
            self.send_header("Content-type", "application/json")
            self.end_headers()

    @staticmethod
    def create_resp_body(value, type, expires_in):
        return "{'access_token':'" + value + "','token_type':'" + type + "','expires_in':" + expires_in + "}";



